### ISSUE TITLE
Use mpmath to compute numerical SVD (for Matrix of Floats)

### ISSUE THREAD

**Issue by cbm755** [sympy/sympy#12277]
If a Matrix has only Floats in it, and user wants the SVD, we could call `mpmath.svd` instead.

I assume (but haven't checked) that mpmath uses a numerically stable algorithm (being primarily a numerical package).  It also does the singular vectors and supports both "full and economy size" SVD.

This should also be done if the matrix had a mix of Floats and Integers.  Perhaps using this logic:
````
is_Float_matrix = (any([x.is_Float for x in A]) and all([x.is_Float or x.is_Integer for x in A]))
````

@siefkenj have you thought about doing this?  We could do this for LU and QR as well.  Should be easier than implementing numerically stable algorithms in SymPy...

**Comment by cbm755** [comment]
@siefkenj I've now tried this [downstream in my Octave package](https://github.com/cbm755/octsympy/pull/766).  It seems to work well!  Let me know what you think.

**Comment by siefkenj** [comment]
I think deferring to `mpmath` for the numerical matrix operations on pure float/integer matrices is a good idea.

```is_NumericalMatrix = all(x.is_Float or x.is_Integer for all x in A)```

should be sufficient since `any([x.is_Float for x in A])` has to run through the entire matrix anyways to return `False`.

Even for matrix multiplication, this could speed things up for purely numerical matrices.  For `ImmutableMatrix` whether it is a numerical type only needs to be computed once.  For `MutableMatrix` it needs to be computed before every such operation...


Does `mpmath` handle sparse matrices?  It is important that dense and sparse versions of the same matrix have the same behavior.

**Comment by cbm755** [comment]
Good point re: multiplication: this is getting more into your `FloatMatrix` idea (which would just be a thin layer of `mpmath.matrix` I suppose).

> should be sufficient

I'm not so sure: I guess someone making an `Integer` `Matrix` will be annoyed to get non-exact calculations done to it: that is why I said it would have at least one Float.

>  It is important that dense and sparse versions of the same matrix have the same behavior.

I haven't looked at `mpmath` for sparse matrices.  I agree with this statement to some extent, but with some caveats:

1.  one should not expect identical numerical results from algorithms on sparse matrices b/c the order of floating point calculations could be different.
2. the algorithms one might run are generally different: e.g., if you ask for the LU factorization of a sparse matrix, you might as a well just make it dense before you start b/c its going to fill-in anyway.   To overly generalize: direct algorithms for dense, iterative algorithms for sparse (`eig` versus `eigs` in Numpy/Matlab).

**Comment by siefkenj** [comment]
My mistake.  I thought `mpmath` automatically did BigInt arithmetic.

**Comment by sylee957** [comment]
Regarding the issue like #17218, I think that mpmath still lacks some obvious ones like RREF, and neither numpy.
http://mpmath.org/doc/current/matrices.html#

And such things may already be too much off-discussion among the developers of numeric libraries, because of the engineering causes.
So some rarely-used numeric things may still need to be implemented here.
